File: python-stuff/credit.py
# TODO


from cs50 import get_int
import math


def main():
    ccNo = get_int("Number: ")

    sumOne = 0
    sumTwo = 0
    currDigit = -1
    prevDigit = -1
    divisor = 1
    cardType = "INVALID\n"

    for i in range(1, 17):
        divisor = divisor * 10

        if currDigit >= 0:
            prevDigit = currDigit

        currDigit = math.floor((ccNo % divisor) / (divisor / 10))
        # print(f"currDigit: {currDigit}")

        if i % 2 == 0:
            doubleDigit = currDigit * 2
            digitVal = 0

            if doubleDigit >= 10:
                firstSf = math.floor((doubleDigit % 100) / 10)
                secondSf = math.floor(doubleDigit % 10)
                digitVal = firstSf + secondSf
            else:
                digitVal = doubleDigit

            # print(f"digitVal: {digitVal}")
            sumTwo += digitVal
        else:
            sumOne += currDigit

        if i == 13 and currDigit == 4:
            cardType = "VISA\n"

        if (i == 15 and currDigit == 3) and (prevDigit == 4 or prevDigit == 7):
            cardType = "AMEX\n"

        if (i == 16 and currDigit == 5) and (prevDigit > 0 and prevDigit < 6):
            cardType = "MASTERCARD\n"

        if i == 16 and currDigit == 4:
            cardType = "VISA\n"

    checkSum = sumOne + sumTwo
    # print(f"sumOne: {sumOne}")
    # print(f"sumTwo: {sumTwo}")
    # print(f"checkSum: {checkSum}")

    if checkSum % 10 != 0:
        print("INVALID\n")
    else:
        print(f"{cardType}\n")


main()

File: C-stuff/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>

int main(void)
{
    // prompt for credit card number
    long ccNumber = get_long("Number: ");

    // initialise variables
    int sumOne = 0;
    int sumTwo = 0;
    int currDigit = -1;
    int prevDigit = -1;
    long divisor = 1;
    string cardType = "INVALID\n"; // pre define cardType as invalid for cards that pass check sum but are not valid

    int i;
    for (i = 1; i <= 16; i++)
    {
        divisor = divisor * 10; // increase divisor each step
        if (currDigit >= 0)
        {
            // if there is a valid currDigit from previous step, set to prevDigit
            prevDigit = currDigit;
        }
        currDigit = (ccNumber % divisor) / (divisor / 10); // get current digit
        if (i % 2 == 0)
        {
            // if current digit is every other digit starting with second-to-last digit, times current digit by 2
            int doubleCurrDigit = currDigit * 2;
            int digitVal;
            if (doubleCurrDigit >= 10)
            {
                // if product is more than single digit separate them and set digitVal to sum of both digits
                int firstSF = (doubleCurrDigit % 100) / 10;
                int secondSF = doubleCurrDigit % 10;
                digitVal = firstSF + secondSF;
            }
            else
            {
                // if product is a single digit, set digitVal to doubleCurrDigit
                digitVal = doubleCurrDigit;
            }
            sumTwo += digitVal;
        }
        else
        {
            // else add to sumOne
            sumOne += currDigit;
        }

        // if 13th digit == 4 then visa
        if (i == 13 && currDigit == 4)
        {
            cardType = "VISA\n";
        }
        // if 15th digit == 3 and prevDigit is a 4 or 7 then american express
        if ((i == 15 && currDigit == 3) && (prevDigit == 4 || prevDigit == 7))
        {
            cardType = "AMEX\n";
        }
        // if 16th digit == 5 and prevDigit is a 1, 2, 3, 4 or 5 then mastercard
        if ((i == 16 && currDigit == 5) && (prevDigit > 0 && prevDigit < 6))
        {
            cardType = "MASTERCARD\n";
        }
        // if 16th digit == 4 then visa
        if (i == 16 && currDigit == 4)
        {
            cardType = "VISA\n";
        }
    }

    // check check sum and print the answer
    int checkSum = sumOne + sumTwo;
    if (checkSum % 10 != 0)
    {
        printf("INVALID\n");
    }
    else
    {
        printf("%s", cardType);
    }
}
